desktop/src/backends/ui.rs

```diff
@@ -408,3 +395,127 @@ impl UiBackend for DesktopUiBackend {
 
     fn close_file_dialog(&mut self) {}
 }
+
+fn load_font_from_file(
+    path: &Path,
+    name: String,
+    index: u32,
+    is_bold: bool,
+    is_italic: bool,
+) -> Result<FontDefinition<'static>> {
+    match std::fs::read(path) {
+        Ok(data) => Ok(FontDefinition::FontFile {
+            name,
+            is_bold,
+            is_italic,
+            data,
+            index,
+        }),
+        Err(e) => Err(anyhow!("Couldn't read font file at {path:?}: {e}")),
+    }
+}
+
+fn load_fontdb_font(name: String, face: &FaceInfo) -> Result<FontDefinition<'static>> {
+    let is_bold = face.weight > fontdb::Weight::NORMAL;
+    let is_italic = face.style != fontdb::Style::Normal;
+
+    match &face.source {
+        fontdb::Source::File(path) => {
+            load_font_from_file(path, name, face.index, is_bold, is_italic)
+        }
+
+        fontdb::Source::Binary(bin) | fontdb::Source::SharedFile(_, bin) => {
+            Ok(FontDefinition::FontFile {
+                name,
+                is_bold,
+                is_italic,
+                data: bin.as_ref().as_ref().to_vec(),
+                index: face.index,
+            })
+        }
+    }
+}
+
+#[cfg(feature = "fontconfig")]
+fn fontconfig_sort_device_fonts(
+    query: &FontQuery,
+    register: &mut dyn FnMut(FontDefinition),
+) -> Vec<FontQuery> {
+    use fontconfig::{FontFormat, Pattern};
+    use std::sync::LazyLock;
+
+    static FONTCONFIG: LazyLock<Option<fontconfig::Fontconfig>> =
+        LazyLock::new(fontconfig::Fontconfig::new);
+
+    let Some(fc) = FONTCONFIG.as_ref() else {
+        return Vec::new();
+    };
+
+    let Ok(family) = std::ffi::CString::new(query.name.as_str()) else {
+        tracing::error!("Cannot sort device fonts, null in font family");
+        return Vec::new();
+    };
+
+    let mut pattern: Pattern<'static> = Pattern::new(fc);
+    pattern.add_string(fontconfig::FC_FAMILY, family.as_c_str());
+
+    if query.is_bold {
+        pattern.add_integer(fontconfig::FC_WEIGHT, fontconfig::FC_WEIGHT_BOLD);
+    }
+    if query.is_italic {
+        pattern.add_integer(fontconfig::FC_SLANT, fontconfig::FC_SLANT_ITALIC);
+    }
+
+    let font_set = pattern.sort_fonts(true);
+    let mut font_queries = Vec::new();
+    for font in font_set.iter() {
+        let is_ttf = font
+            .format()
+            .is_ok_and(|f| matches!(f, FontFormat::TrueType));
```

Comment:
do we want some info log, if our query matches a font but we happen to not be able to load it as it's not TTF?

Reply:
Added